### ISSUE TITLE
Nested language not parsed correctly

### ISSUE THREAD

**Issue by xixiaofinland** [nvim-treesitter/nvim-treesitter#5383]
### Describe the bug

Issue: the parser(s) works differently in Nvim and the playground. 

[This repo](https://github.com/aheber/tree-sitter-sfapex/tree/main) defines three language tree-sitter parsers: `apex` language can include nested `sosl` language (which in turn can include `soql` language), in [grammar.js](https://github.com/aheber/tree-sitter-sfapex/blob/31853f4ca3303919e7715b2517df39330baaf6bf/apex/grammar.js#L12)

I installed the parsers in Nvim, but once it has nested language (e.g.`apex` with nested `soql`), Nvim doesn't understand anymore (screenshot below).

![Screenshot 2023-09-12 at 15 04 24](https://github.com/nvim-treesitter/nvim-treesitter/assets/13655323/c44d0a87-3526-4975-8de0-13c0ff120b05) 

But in the [playground](https://aheber.github.io/tree-sitter-sfapex/playground/) built by the parser author, it works as expected.
![Screenshot 2023-09-12 at 15 23 33](https://github.com/nvim-treesitter/nvim-treesitter/assets/13655323/12b01bbc-ed04-47f3-a43e-cb9858d0a976)

Code:
```
public class hello {
  public hello (){
    Account a = [select id from Account];
  }
}
```

### To Reproduce

1. Point the 3 language parsers to local folder (config below)
2. Make sure the 3 languages are installed by nvim-tree-sitter plugin

```
    local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
    parser_config.apex = {
      install_info = {
        url = "~/projects/tree-sitter-sfapex/apex", -- local path or git repo
        files = { "src/parser.c" },                 -- note that some parsers also require src/scanner.c or src/scanner.cc
        -- optional entries:
        branch = "main",                            -- default branch in case of git repo if different from master
        generate_requires_npm = false,              -- if stand-alone parser without npm dependencies
        requires_generate_from_grammar = false,     -- if folder contains pre-generated src/parser.c
      },
    }

    parser_config.soql = {
      install_info = {
        url = "~/projects/tree-sitter-sfapex/soql", -- local path or git repo
        files = { "src/parser.c" },
        branch = "main",                            -- default branch in case of git repo if different from master
        generate_requires_npm = false,              -- if stand-alone parser without npm dependencies
        requires_generate_from_grammar = false,     -- if folder contains pre-generated src/parser.c
      }
    }

    parser_config.sosl = {
      install_info = {
        url = "~/projects/tree-sitter-sfapex/sosl", -- local path or git repo
        files = { "src/parser.c" },
        branch = "main",                            -- default branch in case of git repo if different from master
        generate_requires_npm = false,              -- if stand-alone parser without npm dependencies
        requires_generate_from_grammar = false,     -- if folder contains pre-generated src/parser.c
      }
    }
```

### Expected behavior

the `InspectTree` should show successful nodes as in the playground.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- OK `tree-sitter` found 0.20.8 (parser generator, only needed for :TSInstallFromGrammar)
- OK `node` found v18.16.0 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: Apple clang version 14.0.0 (clang-1400.0.29.102)
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "21.6.0",
  sysname = "Darwin",
  version = "Darwin Kernel Version 21.6.0: Wed Aug 10 14:25:27 PDT 2022; root:xnu-8020.141.5~2/RELEASE_X86_64"
} ~

Parser/Features         H L F I J
  - apex                ✓ ✓ ✓ . ✓
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - css                 ✓ . ✓ ✓ ✓
  - diff                ✓ . . . .
  - gitcommit           ✓ . . . ✓
  - gitignore           ✓ . . . .
  - html                ✓ ✓ ✓ ✓ ✓
  - java                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - json                ✓ ✓ ✓ ✓ .
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - soql                ✓ . . . .
  - sosl                ✓ . . . .
  - toml                ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓
  - xml                 ✓ . ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.9.1
Build type: Release
LuaJIT 2.1.0-beta3

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.9.1/share/nvim"
```

**Comment by clason** [comment]
Neovim queries are different than upstream queries. Make sure you take this into account.

**Comment by xixiaofinland** [comment]
> Neovim queries are different than upstream queries. Make sure you take this into account.

Do you mean the `scm` queries? This is parser only, queries are not involved IMO.

**Comment by clason** [comment]
Queries are involved when you're talking about injections.

**Comment by xixiaofinland** [comment]
> 

ahh really? That could be the cause! Thanks clason, I will check in details! @clason

**Comment by clason** [comment]
also, you need to specify `requires_generate_from_grammar = true` since the repo doesn't contain any `parser.c`.

**Comment by xixiaofinland** [comment]
> also, you need to specify `requires_generate_from_grammar = true` since the repo doesn't contain any `parser.c`.

I installed `tree-sitter` cli locally and generated three parser.c by myself.

**Comment by xixiaofinland** [comment]
> 

I wonder if it's because the `soql` filetype detects the language syntax from the file content then attempts to get involved? Somehow recall vim filetype does all possible ways to decide if a file is a matching filetype.

I will test around and also try `--clean` setup

**Comment by xixiaofinland** [comment]
> Queries are involved when you're talking about injections.

btw, I think I already got the injection for `soql` correctly in my own [dotfiles](https://github.com/xixiaofinland/dotfiles/blob/main/.config/nvim/after/queries/apex/injections.scm), otherwise those nested `soql` won't show syntax at all

**Comment by xixiaofinland** [comment]
Though I didn't get the solution, I found some facts:

1. The `apex` parser already includes `soql` parsing rules in the compile time. So it won't dynamically call other parsers
2. Whenever nested `soql` appears in `apex` file, the `soql` parser kicks in to also parse the entire file
3. Once I uninstall `soql` parser, the inspectTree nodes are correct (except the soql syntax highlights are gone, as expected).
![Image 007](https://github.com/nvim-treesitter/nvim-treesitter/assets/13655323/248b3d18-8bd0-4ea8-a35f-f07251137b2d)

So the question is: what invokes `soql` parser to kicks in when the nested soql is detected?

@clason any tip?

**Comment by clason** [comment]
I have no idea how that parser works. If it doesn't use injections for nested languages, it's very unusual.

**Comment by xixiaofinland** [comment]
It is the `injections.scm`, if I remove it, the `soql` parser won't kick in anymore. 
```
(query_expression
 (soql_query) @injection.content
      (#set! injection.language "soql"))
```

The apex parser imported the other two parsers like this:
https://github.com/aheber/tree-sitter-sfapex/blob/31853f4ca3303919e7715b2517df39330baaf6bf/apex/grammar.js#L12
https://github.com/aheber/tree-sitter-sfapex/blob/31853f4ca3303919e7715b2517df39330baaf6bf/apex/grammar.js#L12 

I bet this is not the best practice of nvim-tree-sitter, it conflicts with `injections.scm` way of working.


Btw, anything else we need to do to add these parsers into nvim-tree-sitter repo?  We have opened the request [here](https://github.com/nvim-treesitter/nvim-treesitter/issues/2282#issuecomment-1698883867), the parsers are working with well-structured tests, and we are also actively fine-tuning the queries. @clason

**Comment by clason** [comment]
Oh, so you _are_ using an injections query. Sorry, this is a bit hard to follow. 

Yes, nested languages are supposed to be handled with separate parsers and an injections query; that's the tree-sitter (not just nvim-treesitter) way.

What needs to be done to add these parsers is a PR to add them -- and queries that align with the Neovim/nvim-treesitter guidelines (see CONTRIBUTING.md).

**Comment by xixiaofinland** [comment]
> Oh, so you _are_ using an injections query. Sorry, this is a bit hard to follow.
> 
> Yes, nested languages are supposed to be handled with separate parsers and an injections query; that's the tree-sitter (not just nvim-treesitter) way.
> 
> What needs to be done to add these parsers is a PR to add them -- and queries that align with the Neovim/nvim-treesitter guidelines (see CONTRIBUTING.md).

but I use `injections.scm` purely for getting the soql syntax highlight without the need for the soql parser, as the parsing rules are already part of Apex. Can I achieve it?

I will follow the PR, thanks for the advice.

**Comment by clason** [comment]
> but I use injections.scm purely for getting the soql syntax highlight without the need for the soql parser, as the parsing rules are already part of Apex. Can I achieve it?


That makes no sense. injections tell tree-sitter to _use_ the soql parser.

**Comment by xixiaofinland** [comment]
> > but I use injections.scm purely for getting the soql syntax highlight without the need for the soql parser, as the parsing rules are already part of Apex. Can I achieve it?
> 
> That makes no sense. injections tell tree-sitter to _use_ the soql parser.
I see.

but still, the core mystery is unsolved. 
In the `injections.scm`, I clearly defined the scope of `(query_expression (soql_query))`, why would `soql` attempts to parse and syntax highlight the entire file?
